### PR TITLE
Initial WGPU Frontend

### PR DESCRIPTION
**Pull request by zer0x64**
This is an initial front-end using raw `wgpu` instead of SDL.

~~Currently there are performance issues. I will benchmark it a bit to find what's the problem. It can either be `wgpu`, the way I use `wgpu` or the emulator itself.~~ The bottleneck was because of VSYNC, which messed up the timings. Neither `wgpu` nor the way it's used here cause a significant performance hit compared to the emulator itself.

There's also a TODO to address before merging.

Before it is ready to replace `nestadia-gui` for real, I will need to add debugging support, which is far from being trivial, ~~and save files~~.

### CODE REVIEW

File: nestadia-wgpu/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+                    }
+
+                    WindowEvent::KeyboardInput { input, .. } => match input {
+                        // Exit if ESC is pressed
+                        KeyboardInput {
+                            state: ElementState::Pressed,
+                            virtual_keycode: Some(VirtualKeyCode::Escape),
+                            ..
+                        } => {
+                            if let Some(save_data) = state.emulator.get_save_data() {
```
Comment (CBenoit) [inline nestadia-wgpu/src/main.rs:554]: This can certainly be refactored into a function

**Review by CBenoit — APPROVED** [review]
Nice job! I don't know much about all the wgpu part, but this looks quite good to me.

File: nestadia-wgpu/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+        const B = 0x40;
+        const SELECT = 0x20;
+        const START = 0x10;
+        const UP = 0x08;
+        const DOWN = 0x04;
+        const LEFT = 0x02;
+        const RIGHT = 0x01;
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (zer0x64) [reply nestadia-wgpu/src/main.rs:46]: I'll do it in it's own PR because it requires changes in every front-end!

**Comment by zer0x64** [comment]
I just need to move the controller inputs to `input()` and then it's ready to merge I think!

**Pull request merged**